### PR TITLE
Fix Spherize – use SVD, simplify calculations

### PR DESCRIPTION
**Pull request by shntnu**
# Description

Make several updates to `Spherize`

- Use SVD for sphering instead of first computing covariance matrix followed by eigenvalue decomposition. This notebook demonstrates the equivalence [svd-eig.pdf](https://github.com/cytomining/pycytominer/files/12566840/FireShot.Capture.017.-.test.-.pdf).
- Use `epsilon` to regularize, not clip (the former is more standard; unclear why we were doing the latter)
- `PCA-cor` and `ZCA-cor` necessary center the data, so raise an error if centering is `False`
- Handle wide matrices 

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: pycytominer/normalize.py
Comment (gwaybio) [inline pycytominer/normalize.py:165]: Amazing to catch this Shantanu! Thanks for this proposed fix.

I wonder if a simpler solution is to output the numpy array from `spherize()` directly. i.e., the return line at https://github.com/shntnu/pycytominer/blob/33cbb30f8030fddc252d3f470358df289035f977/pycytominer/operations/transform.py#L165

should simply be `return X @ self.W` (or what you have edited to include, I haven't looked yet; basically, we should return the values instead of a DataFrame, which might be a simpler fix. Oh, but maybe this still breaks b/c the columns will be principal components and not `feature_df.columns`?

File: pycytominer/normalize.py
Comment (gwaybio) [inline pycytominer/normalize.py:183]: I like these checks, but they could be frustrating errors for our users. Is there a way to be more specific about what to check if this assertion fails?

File: pycytominer/operations/transform.py
Comment (gwaybio) [inline pycytominer/operations/transform.py:61]:
```suggestion
        if self.method in ["PCA-cor", "ZCA-cor"]:
            assert(self.center), "PCA-cor and ZCA-cor require center=True"
```

This feels more readable to me

File: pycytominer/operations/transform.py
Comment (gwaybio) [inline pycytominer/operations/transform.py:124]: I fear that this would take a while to revisit after we merge, would it take too long to explain?

File: pycytominer/normalize.py
Comment (gwaybio) [reply pycytominer/normalize.py:165]: Oh, I think I understand. We want to use `Spherize()` standalone. I wonder then if it is more pythonic to add an argument like `Spherize(return_numpy=False)` and then set `normalize()` internals to use `return_numpy=True` in `Spherize` internals. What do you think?

File: pycytominer/operations/transform.py
Comment (gwaybio) [inline pycytominer/operations/transform.py:97]: ls there a reason to not set X as `X = StandardScaler(with_mean=True, with_std=False).fit_transform(X)`?

File: pycytominer/operations/transform.py
Comment (gwaybio) [inline pycytominer/operations/transform.py:105]: This error message could be bolstered as well to reduce user frustration. Is there somethign we could recommend a user do to make it work?

File: pycytominer/operations/transform.py
Comment (gwaybio) [inline pycytominer/operations/transform.py:114]: Same comment about these error messages

File: pycytominer/operations/transform.py
Comment (gwaybio) [inline pycytominer/operations/transform.py:129]: will this assertion ever be triggered? In other words, can we test it?

File: tests/test_normalize.py
Comment (gwaybio) [inline tests/test_normalize.py:449]: Can you add a test that enters the assertion? This is a great spot for it!

File: tests/test_operations/test_transform.py
Comment (gwaybio) [inline tests/test_operations/test_transform.py:26]: Can you add a test that enters the assertion? This is a great spot for it!

**Review by gwaybio — CHANGES_REQUESTED** [review]
Glad to see this fix, especially since the tests pass!

I made several inline comments, which we should discuss prior to merging. My general comments include:

1. Please add tests for entry into the assertion that centering must be performed
2. We really should also add testing conditions for the other assertions
3. The other assertion statement errors need more user-friendly (read, less potential for frustration) text. Can we prompt a user on what to fix if they run into one?
4. I'm wondering if an easier fix is to optionally return the numpy array rather than the DataFrame in `Spherize()` itself. 

I made more specific comments for each of these four general comments in line below.

Thanks again!

**Comment by gwaybio** [comment]
Oops! @shntnu - I just saw the "removed the request" and WIP tag. Feel free to consider my comments as preliminary or to ignore them

**Comment by johnarevalo** [comment]
This PR fixes #179

File: pycytominer/operations/transform.py
Comment (johnarevalo) [inline pycytominer/operations/transform.py:71]: This should be X.mean(axis=0) or assert X is pd.DataFrame object. If X is numpy array, then X.mean() will be a scalar(global mean)

Comment (johnarevalo) [reply pycytominer/operations/transform.py:71]: nvm 😅, X.values line below is implicitly assuming X is a dataframe.

File: pycytominer/normalize.py
Comment (shntnu) [reply pycytominer/normalize.py:183]: The purpose of this assertion is really to identify new test fixtures, because this indicates a bug in software. I now prompt the user to file an issue.

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:61]: Thanks. I've made these more readable and also changed them to exceptions

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:97]: Ah because I use `self.mean_centerer` later in `Spherize.transform`

```
if self.center:
    X = self.mean_centerer.transform(X)
```

Makes sense?

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:105]: More explanation added

File: pycytominer/normalize.py
Comment (shntnu) [reply pycytominer/normalize.py:165]: What a great idea! This should work

Comment (shntnu) [reply pycytominer/normalize.py:165]: Works!

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:71]: Right, I do this within `fit`

```
        # Get Numpy representation of the DataFrame
        X = X.values
```

Maybe I should have renamed the variance but keeping it `X` makes it cleaner

Comment (shntnu) [reply pycytominer/operations/transform.py:71]: The new changes have made this outdated so I will resolve

File: tests/test_normalize.py
Comment (shntnu) [reply tests/test_normalize.py:449]: nice

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:122]: Now explained in more detail

File: tests/test_operations/test_transform.py
Comment (shntnu) [reply tests/test_operations/test_transform.py:26]: I do this in `test_normalize_spherize` instead

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:157]: It may be triggered if `np.linalg.svd` ever changes behavior. We cannot test it (easily) because it will involve mocking a version of `np.linalg.svd` that has this behavior.  I use `create_bug_report_message` here, which asks the user to report a bug.

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py]: Turns out this was very likely an incorrect requirement 
I've added this comment in the code for our future selves
```
            # Note: There was previously a requirement r==d otherwise the
            # ZCA transform would not be well defined. However, it later appeared
            # that this requirement was not necessary.
```

File: pycytominer/normalize.py
Comment (shntnu) [reply pycytominer/normalize.py]: I now use `create_bug_report_message`

**Review by gwaybio — COMMENTED** [review]
Thanks for updating this code Shantanu!

I'm going to propose the following to get this merged in ASAP

- @mattsoncam - can you focus on the math side of things? The current tests pass so we can be confident about accuracy, but it would be great for you to read for understanding and provide suggestions on where the documentation can be upgraded.
- @kenibrewer - can you focus on the integration aspects? The added bug report function in `util.py` is neat, but you have your finger on the pulse of the pycytominer roadmap. Can you provide your thoughts here?

I think this can be a relatively quick PR merge, which I will do once we have these two approvals. Sorry about the delays Shantanu, we're right in the middle of me phasing out of PR review in favor of much more knowledgeable/qualified maintainers :)

**Review by kenibrewer — DISMISSED** [review]
Would it be possible to move the new `create_bug_report_message` function into a separate branch and pull request to be reviewed separately? There are a few aspects about the new pattern that merit discussion and I would prefer to keep this PR focused specifically on the spherize/SVD changes. 

EDIT: To clarify. I agree with Greg that the feature has value, but it would be good to have a plan (and some documentation) about when pycytominer wants to "own" the error message as a bug and when we want to consider it data/user error and instead provide instructions on how to fix the problem.

File: pycytominer/operations/transform.py
Comment (MattsonCam) [inline pycytominer/operations/transform.py:64]: This is a good approach. Could also just change self.center, but this may be better for informing the user.

File: pycytominer/operations/transform.py
Comment (MattsonCam) [inline pycytominer/operations/transform.py:114]: Consider mentioning why sphering is not supported in this case.

File: pycytominer/operations/transform.py
Comment (MattsonCam) [inline pycytominer/operations/transform.py:109]: Consider mentioning why the rank should follow these properties

File: pycytominer/operations/transform.py
Comment (MattsonCam) [inline pycytominer/operations/transform.py:63]: Consider adding a source/documentation for why this is required.

**Review by MattsonCam — COMMENTED** [review]
Very detailed pr. I left a few comments to consider. Great job!

File: pycytominer/cyto_utils/util.py
Comment (shntnu) [reply pycytominer/cyto_utils/util.py:368]: @kenibrewer has requested moving this into a new PR, so will do that there

File: pycytominer/normalize.py
Comment (shntnu) [reply pycytominer/normalize.py]: I have changed it to `The number of rows and columns in the normalized dataframe does not match the original dataframe` because it needn't be cells (could be well, for example), and we also check the number of columns (not just rows)

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:63]: Thanks for the nudge

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:64]: Ah yes, good point. I'd favor erroring out in this case.

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:120]: Good catch; `X` is the data matrix. I now cite https://stats.stackexchange.com/q/134282/8494

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:126]: If n > d (long matrix) then Sigma would be dxd -- is that your question?

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:114]: Great catch!

Sphering is indeed supported in this case, but it's just highly unlikely that we'd end up with a perfect linear dependency, so we should not proceed. I have updated the error message. LMK if this suffices.

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:109]: Now added

```
        # Case 1: More features than samples (n < d).
        # If centered (mean of each feature subtracted), one dimension becomes dependent, reducing rank to n - 1.
        # If not centered, the max rank is limited by n, as there can't be more independent vectors than samples.
        # Case 2: More samples than features or equal (n >= d).
        # Here, the max rank is limited by d (number of features), assuming each provides unique information.
```

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:146]: Thanks for the nudge; I've now written it out

**Comment by shntnu** [comment]
> Would it be possible to move the new `create_bug_report_message` function into a separate branch and pull request to be reviewed separately?

Makes sense, but instead, I'll just copy the code snippet here because I bet you'd want to do this more systematically later

<details>

```py
def create_bug_report_message(error_detail, context):
    """
    Constructs an error message with a bug report prompt.

    Parameters
    ----------
    error_detail: str
        Description of the error.
    context: str
        A description of the context for the error.
    Returns
    -------
    str
        Error message with a link to file a bug report.
    """

    bug_report_url = "https://github.com/cytomining/pycytominer/issues"
    return (
        f"{error_detail} This is likely a bug in `{context}`. "
        "Please help us improve our software by filing a bug report at "
        f"{bug_report_url}."
    )


def test_create_bug_report_message():
    error_detail = "Division by zero error occurred."
    context = "calculate_statistics function"
    actual_message = create_bug_report_message(error_detail, context)

    # check that the strings `error_detail` and `context` are in the message
    assert error_detail in actual_message
    assert context in actual_message

    # check that the message contains a link to file a bug report
    assert "https://github.com/cytomining/pycytominer/issues" in actual_message

```

**Comment by shntnu** [comment]
> Very detailed pr. I left a few comments to consider. Great job @shntnu !

Thank you for nudging me to clarify the math further @MattsonCam! I understood my code better after explaining it :D I've resolved all your comments but feel free to reopen if necessary.

**Comment by gwaybio** [comment]
It would be great to get this fix merged in. 

I just chatted with @MattsonCam and he will give a 👍 

@kenibrewer - is this ready from your perspective?

File: pycytominer/operations/transform.py
Comment (MattsonCam) [reply pycytominer/operations/transform.py:200]: The documentation looks great; I didn't know numpy implemented computing the whitening matrix. May also consider providing the numpy link

File: pycytominer/operations/transform.py
Comment (MattsonCam) [reply pycytominer/operations/transform.py:136]: Oh, that makes sense. You are making sure there is no bug when implementing numpy's svd function. If you're returning the full matrices, then Sigma should be (nxd), even if it's sparse, from my understanding

File: pycytominer/operations/transform.py
Comment (MattsonCam) [reply pycytominer/operations/transform.py:130]: Ahh, I see, thanks for the resource

File: pycytominer/operations/transform.py
Comment (MattsonCam) [reply pycytominer/operations/transform.py]: This looks like it will suffice!

File: pycytominer/operations/transform.py
Comment (MattsonCam) [reply pycytominer/operations/transform.py]: That looks good!

File: pycytominer/operations/transform.py
Comment (MattsonCam) [reply pycytominer/operations/transform.py]: Saw the comment, well done

**Comment by MattsonCam** [comment]
> > Very detailed pr. I left a few comments to consider. Great job @shntnu !
> 
> Thank you for nudging me to clarify the math further @MattsonCam! I understood my code better after explaining it :D I've resolved all your comments but feel free to reopen if necessary.

Absolutely! Thank you for addressing them @shntnu! I've finished addressing your response comments, but didn't feel it was imperative to reopen any of them. 

> It would be great to get this fix merged in.
> 
> I just chatted with @MattsonCam and he will give a 👍 when he gets a chance
> 
> @kenibrewer - is this ready from your perspective?

I give the 👍

**Review by MattsonCam — APPROVED** [review]

File: pycytominer/operations/transform.py
Comment (shntnu) [reply pycytominer/operations/transform.py:202]: Ah, `numpy` does not implement computing the whitening matrix. Maybe you are referring to the comment "In NumPy, this is implemented as"? That's referring to`W = (np.linalg.inv(S) @ Vt) * np.sqrt(n - 1)` but you still need to compute `Vt`, plus computing `np.linalg.inv(S)` explicitly is too slow/memory intensive. I have now updated that comment.

**Comment by shntnu** [comment]
> I give the 👍

Hooray! 

Over to one of you to merge once the tests pass (I made a documentation edit)

**Comment by gwaybio** [comment]
this looks good to me @shntnu - I am going to merge this in now. Thanks for this contribution and for your patience as we reviewed! 🎉

**Review by gwaybio — APPROVED** [review]

**Comment by gwaybio** [comment]
@kenibrewer - thanks for your initial review. Given @MattsonCam 's approval (and your initial review focused on the bug report document now described in #365 ) - I will merge this in now to avoid additional delays. Thanks!

**Comment by kenibrewer** [comment]
Sorry I missed the pings here. I had modified my github notifications settings incorrectly and didn't get the usual notification emails I rely on. 

Everything looks good to me.

**Comment by gwaybio** [comment]
No worries!

**Comment by shntnu** [comment]
Hooray! Thank you all.

**Pull request merged**
